File: 19.Backend Scaling and Performance Engineering - Part-2/code/typescript/example_4.ts
import express from 'express';
import client from 'prom-client';

const app = express();

const requestDuration = new client.Histogram({
    name: 'http_request_duration_seconds',
    help: 'Duration of HTTP requests',
    labelNames: ['method', 'path', 'status'],
});

// Register it
client.register.registerMetric(requestDuration);

app.get('/metrics', async (req, res) => {
    res.set('Content-Type', client.register.contentType);
    res.send(await client.register.metrics());
});

File: 19.Backend Scaling and Performance Engineering - Part-2/code/rust/example_4.rs
use actix_web::{get, HttpResponse, Responder};
use prometheus::{HistogramVec, HistogramOpts, register_histogram_vec};
use lazy_static::lazy_static;

lazy_static! {
    static ref REQUEST_DURATION: HistogramVec = register_histogram_vec!(
        HistogramOpts::new("http_request_duration_seconds", "Duration of HTTP requests"),
        &["method", "path", "status"]
    ).unwrap();
}

#[get("/metrics")]
async fn metrics() -> impl Responder {
    use prometheus::Encoder;
    let encoder = prometheus::TextEncoder::new();
    let mut buffer = vec![];
    encoder.encode(&prometheus::gather(), &mut buffer).unwrap();
    HttpResponse::Ok().body(buffer)
}
